prov/efa/src/rxr/rxr_ep.c

```diff
@@ -133,50 +133,120 @@ struct rxr_rx_entry *rxr_ep_get_rx_entry(struct rxr_ep *ep,
-		rxr_pkt_entry_copy(ep, unexp_entry, pkt_entry,
+		rxr_pkt_entry_copy(ep, unexp_pkt_entry, pkt_entry,
 				   RXR_PKT_ENTRY_UNEXP);
 		rxr_pkt_entry_release_rx(ep, pkt_entry);
 	} else {
-		unexp_entry = pkt_entry;
+		unexp_pkt_entry = pkt_entry;
 	}
 
-	rts_pkt = rxr_get_rts_hdr(unexp_entry->pkt);
+	return unexp_pkt_entry;
+}
 
-	if (rts_pkt->flags & RXR_TAGGED)
-		op = ofi_op_tagged;
-	else
-		op = ofi_op_msg;
+struct rxr_rx_entry *rxr_ep_alloc_unexp_rx_entry_for_rts(struct rxr_ep *ep,
+							 struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_pkt_entry *unexp_entry;
+	struct rxr_rts_hdr *rts_hdr;
+
+	unexp_entry = rxr_ep_get_unexp_pkt_entry(ep, pkt_entry);
+	if (OFI_UNLIKELY(!unexp_entry))
+		return NULL;
 
-	rx_entry = rxr_ep_get_rx_entry(ep, NULL, 0, rts_pkt->tag, ~0, NULL,
-				       unexp_entry->addr, op, 0);
+	rx_entry = rxr_ep_get_rx_entry(ep, NULL, 0, 0, ~0, NULL,
+				       unexp_entry->addr, ofi_op_msg, 0);
 	if (OFI_UNLIKELY(!rx_entry))
 		return NULL;
 
-	rx_entry->state = RXR_RX_UNEXP;
-	rx_entry->total_len = rts_pkt->data_len;
-	rx_entry->rxr_flags = rts_pkt->flags;
-	rx_entry->unexp_rts_pkt = unexp_entry;
+	rts_hdr = rxr_get_rts_hdr(unexp_entry->pkt);
+	if (rts_hdr->flags & RXR_TAGGED)
+		rx_entry->op = ofi_op_tagged;
+	else
+		rx_entry->op = ofi_op_msg;
 
-	if (op == ofi_op_tagged)
+	rx_entry->tag = rts_hdr->tag;
+	rx_entry->rxr_flags = rts_hdr->flags;
+	rx_entry->total_len = rts_hdr->data_len;
+	if (rx_entry->op == ofi_op_tagged)
 		dlist_insert_tail(&rx_entry->entry, &ep->rx_unexp_tagged_list);
 	else
 		dlist_insert_tail(&rx_entry->entry, &ep->rx_unexp_list);
 
+	rx_entry->state = RXR_RX_UNEXP;
+	rx_entry->unexp_pkt = unexp_entry;
+
+	return rx_entry;
+}
+
+struct rxr_rx_entry *rxr_ep_alloc_unexp_rx_entry_for_msgrtm(struct rxr_ep *ep,
+							    struct rxr_pkt_entry **pkt_entry)
+{
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_pkt_entry *unexp_pkt_entry;
+
+	unexp_pkt_entry = rxr_ep_get_unexp_pkt_entry(ep, *pkt_entry);
+	if (OFI_UNLIKELY(!unexp_pkt_entry))
+		return NULL;
+
+	if (unexp_pkt_entry != *pkt_entry)
+		*pkt_entry = unexp_pkt_entry;
+
+	rx_entry = rxr_ep_get_rx_entry(ep, NULL, 0, 0, ~0, NULL,
+				       unexp_pkt_entry->addr, ofi_op_msg, 0);
+	if (OFI_UNLIKELY(!rx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_CQ,
+			"RX entries exhausted.\n");
+		efa_eq_write_error(&ep->util_ep, FI_ENOBUFS, -FI_ENOBUFS);
+		return NULL;
+	}
+
+	rx_entry->rxr_flags = 0;
+	rxr_pkt_read_req_hdr(*pkt_entry, rx_entry);
+	rx_entry->state = RXR_RX_UNEXP;
+	rx_entry->unexp_pkt = unexp_pkt_entry;
+	dlist_insert_tail(&rx_entry->entry, &ep->rx_unexp_list);
+	return rx_entry;
+}
+
+struct rxr_rx_entry *rxr_ep_alloc_unexp_rx_entry_for_tagrtm(struct rxr_ep *ep,
+							    struct rxr_pkt_entry **pkt_entry)
+{
+	uint64_t tag;
+	struct rxr_rx_entry *rx_entry;
+	struct rxr_pkt_entry *unexp_pkt_entry;
+
+	unexp_pkt_entry = rxr_ep_get_unexp_pkt_entry(ep, *pkt_entry);
+	if (OFI_UNLIKELY(!unexp_pkt_entry))
+		return NULL;
```

Comment:
That's not what I'm seeing -- all three of these callers do nothing or only call FI_WARN with the return code:

```
   1    670  src/rxr/rxr_cq.c <<rxr_cq_proc_pending_items_in_recvwin>>
             ret = rxr_pkt_proc_rtm(ep, pending_pkt);
   2    562  src/rxr/rxr_pkt_type_req.c <<rxr_pkt_handle_rtm_recv>>
             rxr_pkt_proc_rtm(ep, pkt_entry);
   3    596  src/rxr/rxr_pkt_type_req.c <<rxr_pkt_handle_rtm_recv>>
             ret = rxr_pkt_proc_rtm(ep, pkt_entry);
```
Is this in a later patch?

Reply:
rxr_pkt_proc_rtm calls rxr_pkt_proc_tagrtm()/ rxr_pkt_proc_msgrtm(), which calls rxr_pkt_get_tagrtm_rx_entry()/rxr_pkt_get_msgrtm_rx_entry()

rxr_pkt_get_tagrtm_rx_entry() will write to EQ if rx_entry was not allocated.

Reply:
For that case, sure, but in the unexpected path (where this comment is) those errors are not handled.

Reply:
I see your point. Added call to FI_WARN and efa_eq_write_err in rxr_ep_alloc_unexp_rx_entry_for_msgrtm() and rxr_ep_alloc_unexp_rx_entry_for_tagrtm(), which is how rxr_ep_alloc_unexp_rx_entry_for_rts() was implemented.

Reply:
It'd be cleaner to change this so that `rxr_pkt_get_msgrtm_rx_entry` (and it's tag equivalent) to call `efa_eq_write_error` after the if/else block.

Also, you're writing two errors when it cannot get a pkt_entry so you can remove the eq error write in `rxr_ep_alloc_unexp_rx_entry_for_msgrtm`.

Reply:
rxr_pkt_get_msgrtm_rx_entry() and rxr_pkt_get_tagrtm_rx_entry() did write eq error (line 416 and 445 of rxr_pkt_type_req.c) .

If removed write_eq_error of packet entry but kept the warning in rxr_ep_alloc_unexp_rx_entry_for_tagrtm().

Reply:
Yup, what is here is correct. I was just pointing out that writing it this way and only having one spot where you write the error is clearer:

```
static
struct rxr_rx_entry *rxr_pkt_get_msgrtm_rx_entry(struct rxr_ep *ep,
                                                 struct rxr_pkt_entry **pkt_entry_ptr)
{
      if (OFI_UNLIKELY(!match)) {
                rx_entry = rxr_ep_alloc_unexp_rx_entry_for_msgrtm(ep, pkt_entry_ptr);
        } else {
                rx_entry = rxr_pkt_get_rtm_matched_rx_entry(ep, match, *pkt_entry_ptr);
        }

        if (OFI_UNLIKELY(!rx_entry)) {
                FI_WARN(&rxr_prov, FI_LOG_CQ, "RX entries exhausted.\n");
                efa_eq_write_error(&ep->util_ep, FI_ENOBUFS, -FI_ENOBUFS);
                return NULL;
        }
}
```